### ISSUE TITLE
Spike: Q64/K64 routed attention on SM120 with precomputed DSA routes

### ISSUE THREAD

**Issue by Boffee** [Boffee/piper-kernels#52]
## Goal

Determine whether irregular routed attention at a nominal 40% video-K budget can execute efficiently enough on SM120 to justify DSA router and Piper Engine integration.

This issue is a kernel feasibility spike, not an end-to-end MiniMax integration.

## Research handoff

The selected research baseline is pinned at [`piper-research@9cf54b7`](https://github.com/Boffee/piper-research/commit/9cf54b75e10ab6fa8b26d1009ef9a10b666eef0c).

- [Production handoff](https://github.com/Boffee/piper-research/blob/9cf54b75e10ab6fa8b26d1009ef9a10b666eef0c/projects/sparse-video-attention/production-handoff.md)
- [Plain-PyTorch semantic reference](https://github.com/Boffee/piper-research/blob/9cf54b75e10ab6fa8b26d1009ef9a10b666eef0c/projects/sparse-video-attention/dsa_reference.py)
- [Exact experiment configuration and results](https://github.com/Boffee/piper-research/blob/9cf54b75e10ab6fa8b26d1009ef9a10b666eef0c/projects/sparse-video-attention/results.json)

The first production candidate uses spatial `1x8x8` Q/K tiles, adaptive `20%/40%/60%` K budgets per head at an exact nominal 40% mean, and fresh DSA routes for every sparse attention invocation. This spike receives routes as inputs and does not compute them.

## Initial operator scope

Use a generic block-major contract equivalent to:

```text
query_blocks       BF16 [B, H, Q_blocks, 64, D]
key_blocks         BF16 [B, H, K_blocks, 64, D]
value_blocks       BF16 [B, H, K_blocks, 64, Dv]
prefix_key         BF16 [B, H, P, D]
prefix_value       BF16 [B, H, P, Dv]
routes             INT32[B, H, Q_blocks, max_keep]
keep_blocks        INT32[H]
scale              FP32 scalar

output             BF16 [B, H, Q_blocks, 64, Dv]
```

Initial specialization:

- batch size 1;
- 56 heads;
- head dimension 128;
- noncausal Q64/K64;
- 1,036 video Q/K blocks;
- 207, 414, or 622 selected K blocks per head;
- ragged 591-token dense prefix;
- precomputed, unique, sorted route indices; and
- exact BF16 routed-attention semantics with float32 softmax accumulation.

Every video Q block attends to all rows from its selected video K blocks plus every real prefix row in one softmax. Route identities vary independently across heads and Q blocks.

## Deliverables

- [ ] Port a portable routed-attention reference and focused tests into this repository. Do not add a runtime dependency on `piper-research`.
- [ ] Define the generic routed-attention API and validation behavior.
- [ ] Implement the narrow SM120 routed-attention kernel with precomputed routes.
- [ ] Add deterministic synthetic route generation matching the real dimensions and 20/40/60 budgets.
- [ ] Optionally validate one ignored local artifact containing real captured routes; do not commit the full route tensor.
- [ ] Benchmark against dense Piper Attention after compilation.
- [ ] Report kernel-only time, route-metadata handling time, and their combined time.
- [ ] Quantify model-specific spatial layout conversion separately before any Engine integration; it must not become part of the generic kernel API.

## Correctness gates

- [ ] All-K routes agree with dense attention under identical arithmetic.
- [ ] Different per-head keep counts execute the requested amount of work.
- [ ] Routes can differ independently across Q blocks.
- [ ] Only active route entries are consumed; padded route entries are ignored.
- [ ] Active route entries are unique, sorted, and range checked by the portable contract.
- [ ] Prefix lengths below, equal to, and above one K block are correct.
- [ ] A ragged 591-token prefix assigns no probability to padding.
- [ ] Spatial boundary padding is either rejected by the initial specialization or masked correctly.
- [ ] Output agrees with the pinned PyTorch reference within arithmetic-appropriate tolerances.

## Benchmark and decision gate

Use both deterministic sorted-random routes and, when available locally, a real captured route. Report absolute latency and the ratio to dense Piper Attention for the same Q/K/V arithmetic and shape.

The provisional continuation target is combined precomputed-route handling plus sparse attention at approximately 60-65% or less of dense attention time. This is a project go/no-go target, not an existing measurement. A result near 80-90% leaves too little room for DSA routing and Engine overhead.

Research FlexAttention timings are quality-harness diagnostics and must not be used as production performance baselines.

## Out of scope for this spike

- DSA score computation or top-K routing kernel;
- 20/40/60 calibration;
- MiniMax-H3 token-layout integration;
- diffusion-call or generation state;
- Q4 or K4 packing;
- end-to-end video generation; and
- default production dispatch.

If the spike passes, follow-up work should cover the DSA router, exact-BF16 versus route-aware quantized Piper arithmetic, and generation-scoped Piper Engine integration.
